### PR TITLE
Add Preferred Events and seperate Personal Recods Route

### PR DESCRIPTION
**Pull request by FinnIckler**
This is needed for the Registration Frontend.

I moved the personal records back out of the `me` route because I started putting preferred events into it, but then I felt like it's getting bloated. I also moved the require_user! method out of the `competition_controller` into the `api_controller` because we now need it everywhere. 
Technically you can always get personal records using the `persons/:wca_id/` route, but I feel like we should keep that separate. Thoughts?

### CODE REVIEW

**Comment by dunkOnIT** [comment]
Personally I favour having more specific routes than bloated general ones. I'd say we add a dedicated `/results/me` or `/results/{wca-id}` route that only returns what is currently in the `rankings` key of `/users/me`

File: WcaOnRails/app/controllers/api/v0/users_controller.rb [hunk 1]
```diff
@@ -2,14 +2,8 @@
...
-    if current_user
-      if stale?(current_user)
-        # Also include the users current prs so we can handle qualifications on the Frontend
-        show_user(current_user, show_rankings: true)
-      end
-    else
-      render status: :unauthorized, json: { error: I18n.t('api.login_message') }
-    end
+    require_user!
+    show_user(current_user)
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/api/v0/users_controller.rb:6]: Where did the `stale?` check go?

File: WcaOnRails/app/controllers/api/v0/users_controller.rb [hunk 2]
```diff
@@ -29,37 +23,43 @@ def show_user_by_wca_id
...
+    require_user!
+    if stale?(current_user)
+      render json: current_user.permissions
     end
   end
 
+  def personal_records
+    require_user!
+    return render json: { single: [], average: [] } unless current_user.wca_id.present?
+    person = Person.includes(:user, :ranksSingle, :ranksAverage).find_by_wca_id!(current_user.wca_id)
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/api/v0/users_controller.rb:35]: Where does the `:user` in the includes come from? As far as I can tell, `ranksSingle` and `ranksAverage` should be enough to compute the return value below.

File: WcaOnRails/app/controllers/api/v0/users_controller.rb [hunk 3]
```diff
@@ -29,37 +23,43 @@ def show_user_by_wca_id
...
+    render json: { single: person.ranksSingle, average: person.ranksAverage }
+  end
+
+  def preferred_events
+    require_user!
+    preferred_events = Rails.cache.fetch("#{current_user.id}-preferred", expires_in: 24.hours) do
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/api/v0/users_controller.rb:41]: Rails cache key doesn't say what is being preferred. Preferred... Delegates? Preferred... home cooking recipes?

File: WcaOnRails/app/controllers/api/v0/users_controller.rb [hunk 4]
```diff
@@ -29,37 +23,43 @@ def show_user_by_wca_id
...
   def bookmarked_competitions
-    if current_user
-      bookmarked_competitions = Rails.cache.fetch("#{current_user.id}-bookmarked", expires_in: 60.minutes) do
-        current_user.competitions_bookmarked.pluck(:competition_id)
-      end
-      render json: bookmarked_competitions
-    else
-      render status: :unauthorized, json: { error: I18n.t('api.login_message') }
+    require_user!
+    bookmarked_competitions = Rails.cache.fetch("#{current_user.id}-bookmarked", expires_in: 60.minutes) do
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/api/v0/users_controller.rb:49]: Again, please specify _what_ is being bookmarked

File: WcaOnRails/app/controllers/api/v0/users_controller.rb [hunk 5]
```diff
@@ -29,37 +23,43 @@ def show_user_by_wca_id
...
-      render status: :unauthorized, json: { error: I18n.t('api.login_message') }
-    end
+    require_user!
+    render json: { status: "ok" }
   end
 
   private
 
-    def show_user(user, show_rankings: false)
+    def show_user(user)
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/api/v0/users_controller.rb:62]: Please do not remove the `show_rankings` option to maintain backwards compatibility.

File: WcaOnRails/spec/controllers/api_users_controller_spec.rb [hunk 6]
```diff
@@ -78,7 +78,6 @@
       expect(response.status).to eq 200
       json = JSON.parse(response.body)
       expect(json["user"]).to eq competed_user.as_json
-      expect(json.key?("rankings")).to eq true
```
Comment (gregorbg) [inline WcaOnRails/spec/controllers/api_users_controller_spec.rb:81]: This test case should not disappear. We might have external API users that rely on it.

**Review by gregorbg — CHANGES_REQUESTED** [review]

File: WcaOnRails/app/controllers/api/v0/users_controller.rb (hunk 1 above)
Comment (FinnIckler) [reply WcaOnRails/app/controllers/api/v0/users_controller.rb:6]: lost in the copy paste

File: WcaOnRails/app/controllers/api/v0/users_controller.rb (hunk 5 above)
Comment (FinnIckler) [reply WcaOnRails/app/controllers/api/v0/users_controller.rb:62]: I added this like 2 mohts ago so I don't expect anyone to actually use it, but I'll keep it in

File: WcaOnRails/app/controllers/api/v0/users_controller.rb (hunk 2 above)
Comment (FinnIckler) [reply WcaOnRails/app/controllers/api/v0/users_controller.rb:35]: it comes from the way it's done in other places of the code where the ranksSingle and ranksAverage are queried, but I guess it's not needed here

**Review by gregorbg — APPROVED** [review]

**Pull request merged**